common/msgBus.js

```diff
@@ -0,0 +1,155 @@
+
+        //Query subscribers
+        this._querySubs = {};
+        
+        //Ids of sended but not answerd queries
+        this._pendingQuerys = new Map();
+    }
+
+    //Creates connection for Redis
+    init() {
+        return new Promise(async (resolve, reject) => {
+
+            let connection = require("../redis-connection.json")
+
+            let config = {
+                port: connection.port,
+                host: connection.host,
+                auth: connection.auth,
+                scope: connection.scopes.service
+            };
+
+            this.nrp = new NRP(config);
+
+            await this.nrp.on(this._hostname, (payload) => this.finishQuery(payload))
+            this.nrp.on("broadcast", (payload) => this.notifyLocalSubscribers(payload), () => resolve());
+        })
+    }
+
+    //Parameters: callBack - function to be subscribed
+    //Subscribes a function which will be invoked on broadcast
+    //Returns Promise<void>
+    async subscribeBroadcast(callBack) {
+        this._broadcastSubs.push(callBack);
+    }
+
+    //Parameters: subject - the purpose of the query, callBack - function to be subscribed
+    //Subscribes a function which will be invoked on query
+    //Returns Promise<void>
+    async subscribeQuery(subject, callBack) {
+        if (!this._querySubs.hasOwnProperty(subject)) {
+            this._querySubs[subject] = [];
+        }
+
+        this._querySubs[subject].push(callBack);
+    }
+
+    //Parameters: payload - object to send
+    //Sends the payload to all broadcast subscribers
+    //Returns Promise<void>
+    async broadcast(payload) {
+        await this.nrp.emit("broadcast", payload);
+    }
+
+    //Parameters: subject - the purpose of the query, payload - object to send
+    //Sends the payload to all subscribers to the specific query
+    //Returns Promise<Object> - received answer
+    query(subject, payload) {
+        let promise =  new Promise((resolve, reject) => {
+            this.createQueryPackage(subject, payload, resolve)
+                .then((queryPackage) => this.nrp.emit("broadcast", queryPackage))
+        });
+
+        return promise;
+    }
+
+    async createQueryPackage(subject, payload, resolve) {
+        let id = await MsgBus.generateGuid();
+
+        this._pendingQuerys.set(id, resolve);
+
+        let queryPackage = {
+            "query": subject,
+            "payload": payload,
+            "query-id": id,
+            "node": this._hostname
+        };
+
+        return queryPackage;
+    }
+
+    async finishQuery(resultPackage) {
+        let result = resultPackage.result;
+        let id = resultPackage["query-id"];
+
+        let resolve = this._pendingQuerys.get(id);
+
+        resolve(result);
+
+        this._pendingQuerys.delete(id);
+    }
+
+    async notifyLocalSubscribers(payload) {
+        let isQuery = payload.hasOwnProperty("query-id");
+
+        if (isQuery) {
+            await this.notifyLocalQuerySubscribers(payload);
```

Comment:
With await you block execution. It will be great if you notify all subscribers at once